### PR TITLE
IEEE-193: create pending order table

### PR DESCRIPTION
**Pull request by EthanHugh**
## Overview

- Resolves #IEEE-193
- Created a table for pending orders in the admin page to allow them to complete orders

## Unit Tests Created

- Test to see if the table is rendered properly
- Test to see if the Complete Order button works properly (currently broken)

## Steps to QA

- Go to any team overview from an admin site (http://localhost:3000/teams/abcde) and view the table

## Issues

- Trying to click on an MUI checkbox never works

### CODE REVIEW

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+            orderItem.requested_quantity.toString();
+        orderInitalValues[`${orderItem.id}-checkbox`] = false;
+    });
+    return orderInitalValues;
+};
+
+const convertToDateTime = (dateString: string) => {
+    const date = new Date(dateString);
+    const formattedDateString =
+        ("0" + date.getUTCHours()).slice(-2) +
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx]: reminder to change this to Local time!
you can use date.toLocaleString() or a combination of date.toLocaleDate() and date.toLocaleTime()

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        <Container
+            className={styles.tableContainer}
+            maxWidth={false}
+            disableGutters={true}
+        >
+            {orders.length > 0 && (
+                <GeneralOrderTitle
+                    title="Requested Items"
+                    isVisible={true}
+                    toggleVisibility={() => alert("show/hide orders")}
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx]: could you implement a simple show/hide functionality here
you can use React's useState hook to set/update state visibility

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+                                                    Qty granted
+                                                </TableCell>
+                                                <TableCell
+                                                    className={`${styles.width1} ${styles.noWrap}`}
+                                                >
+                                                    Time
+                                                </TableCell>
+                                                <TableCell>
+                                                    <Checkbox
+                                                        color="primary"
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx]: remove the margin here, instead can you do `<TableCell align="center">`, that will center your checkbox

File: hackathon_site/dashboard/frontend/src/pages/TeamDetail/TeamDetail.tsx [hunk 4]
```diff
@@ -14,6 +14,7 @@ import {
     TableCell,
 } from "@material-ui/core";
 import Typography from "@material-ui/core/Typography";
+import TeamPendingOrderTable from "../../components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable";
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/pages/TeamDetail/TeamDetail.tsx]: absolute import here

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        ("0" + (date.getUTCMonth() + 1)).slice(-2) +
+        "/" +
+        date.getUTCFullYear();
+    return formattedDateString;
+};
+
+export const TeamPendingOrderTable = () => {
+    const orders = mockPendingOrders;
+    const hardware = mockHardware;
+
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx:60]: could you create a submit order function that when you click submit, it will log the form values from the table

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+                                        </Button>
+                                    </Grid>
+                                    <Grid item>
+                                        <Button
+                                            color="primary"
+                                            variant="contained"
+                                            type="submit"
+                                            disableElevation
+                                            onClick={() => {
+                                                alert(
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx]: can you move this to an external handleSubmit function and pass it into the Formik component

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx (hunk 5 above)
Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx:60]: see comments below

**Review by YuYing-Liang — CHANGES_REQUESTED** [review]
Nice start Ethan!

Left some comments on functionality. You'll need tests for your table, something like making sure every order and its order items shows up properly in the table.

**Comment by YuYing-Liang** [comment]
And a reminder to remove any unused imports (you can either go to the problems tab in pycharm to get this info or it will be logged in your react server terminal)

File: hackathon_site/dashboard/frontend/src/slices/ui/uiSlice.ts [hunk 7]
```diff
@@ -12,6 +12,9 @@ interface UIInitialState {
     inventory: {
         isProductOverviewVisible: boolean;
     };
+    team_detail: {
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/slices/ui/uiSlice.ts]: oh u didn't need to do this, I meant local show/hide with react useState because the TeamDetail page is going to be changing a lot so we don't want global state setting for this.

If you do set the state here, when you navigate away from the page it'll maintain the show/hide state. But if you click on a new team, you'll proabably want to see all their orders expanded which is why we don't need global state here

Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/slices/ui/uiSlice.ts]: with the useState hook the code is simply:
```
const [visibility, setVisibility] = useState(false)
// toggle visibility
setVisibility(!visibility)
```

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+                                                                        ?.name
+                                                                }
+                                                            </TableCell>
+                                                            <TableCell>
+                                                                {
+                                                                    hardware[row.id - 1]
+                                                                        ?.model_number
+                                                                }
+                                                            </TableCell>
+                                                            <TableCell>
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx:186]: why does prettier do this ughhhh so much unnecessary whitespace

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import { within, waitFor } from "testing/utils";
+import { fireEvent } from "@testing-library/react";
+
+describe("team pending order table", () => {
+    test("renders team pending order table", () => {
+        const { getByTestId } = render(<TeamPendingOrderTable />);
+        expect(screen.getByText("Requested Items")).toBeInTheDocument();
+
+        //loop through all pending orders
+        for (let i = 0; i < mockPendingOrdersInTable.length; i++) {
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx]: just fyi you can do `mockPendingOrdersInTable.forEach((currentOrder, idx) => {})` because it's an array, most JS developers don't use normal for loops anymore bc the forEach() function is so much cleaner

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+    let orderInitalValues: Record<string, string | boolean> = {};
+    request.forEach((orderItem) => {
+        orderInitalValues[`${orderItem.id}-quantity`] =
+            orderItem.quantityGranted.toString();
+        orderInitalValues[`${orderItem.id}-checkbox`] = false;
+    });
+    return orderInitalValues;
+};
+
+const convertToDateTime = (dateString: string) => {
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx]: not used, remove pls

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx [hunk 11]
```diff
@@ -0,0 +1,71 @@
+import React from "react";
+import { getByTestId, render, screen } from "testing/utils";
+import { mockHardware, mockPendingOrdersInTable } from "testing/mockData";
+import TeamPendingOrderTable from "components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable";
+import { within, waitFor } from "testing/utils";
+import { fireEvent } from "@testing-library/react";
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx]: pls export fireEvent from testing/utils please and you have 2 imports from testing/utils so merge those together

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+            <MenuItem key={i} role="quantity" value={i.toString()}>
+                {i}
+            </MenuItem>
+        );
+    }
+
+    return entry;
+};
+
+const setInitialValues = (
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx:41]: typo in Initial

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+        for (let i = 0; i < mockPendingOrdersInTable.length; i++) {
+            const currentOrder = mockPendingOrdersInTable[i];
+            const submitButton = getByTestId(`complete-button-${currentOrder.id}`);
+            fireEvent.click(submitButton);
+            await waitFor(() => {
+                for (let j = 0; j < currentOrder.hardwareInTableRow.length; j++) {
+                    const currentRow = currentOrder.hardwareInTableRow[j];
+                    expect(
+                        screen.getByText(
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx]: ah so this doesn't work because you display an alert and the alert doesn't show up on the DOM.
There's no good way to test what shows up on the screen for this function. I originally was thinking you would pass in a function into the table and you can test it that way but we'll have to change that anyways when we link it to the API.

So yeah just remove this test, we'll implement the tests later ... my bad :(

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+                        )
+                    ).toBeInTheDocument();
+                    expect(
+                        screen.getByText(
+                            `"${mockHardware[currentRow.id - 1].id}-checkbox": false`
+                        )
+                    ).toBeInTheDocument();
+                }
+            });
+        }
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx:124]: a test you could add tho is pressing the "All" button will change the dropdown value

Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx:124]: you can also test whether the checkbox actually checks the box (but that's a bit more complicated, I'll leave that as optional)

**Review by YuYing-Liang — COMMENTED** [review]
Good work Ethan!

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx (hunk 8 above)
Comment (EthanHugh) [reply hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx:186]: haha true

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx (hunk 12 above)
Comment (EthanHugh) [reply hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx:41]: I think that's how I spelt it already

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx (hunk 13 above)
Comment (EthanHugh) [reply hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx]: np!

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx (hunk 14 above)
Comment (EthanHugh) [reply hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx:124]: > a test you could add tho is pressing the "All" button will change the dropdown value

should i set the default value to 0 then, so that I can test how it changes?

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx (hunk 12 above)
Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx:41]: I meant line 44

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx (hunk 14 above)
Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx:124]: Try to go about it how the user would:

- when table is rendered dropdown is at max value
- User changes dropdown to lower value
- user clicks "All" and it should update the dropdown value

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+            orderItem.quantityGranted.toString();
+        orderInitialValues[`${orderItem.id}-checkbox`] = false;
+    });
+    return orderInitialValues;
+};
+
+export const TeamPendingOrderTable = () => {
+    const orders = mockPendingOrdersInTable;
+    const hardware = mockHardware;
+    const [visibility, setVisibility] = useState(false);
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx]: initialize to true so all the orders are visible

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+    const toggleVisibility = () => {
+        // toggle visibility
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx]: you can remove the comment because your code is very clear, no need to describe :)

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+                                                    </TableCell>
+                                                    <TableCell
+                                                        className={`${styles.width6} ${styles.noWrap}`}
+                                                    >
+                                                        Qty granted
+                                                    </TableCell>
+                                                    <TableCell
+                                                        className={`${styles.width1} ${styles.noWrap}`}
+                                                    >
+                                                        <Checkbox color="primary" />
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx]: could you add a setFieldValue call here to check all the boxes: 
image

you should also write a test for this because it's custom functionality

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+                expect(
+                    within(
+                        getByTestId(`table-${currentOrder.id}-${currentRow.id}`)
+                    ).getByText(`${mockHardware[currentRow.id - 1].manufacturer}`)
+                ).toBeInTheDocument();
+            });
+        });
+    });
+
+    test("All button changes the value correctly", () => {
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx]: nicceee

**Review by YuYing-Liang — COMMENTED** [review]
some minor things, great job overall!

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+                expect(
+                    within(
+                        getByTestId(`table-${currentOrder.id}-${currentRow.id}`)
+                    ).getByTestId(`select`)
+                ).toHaveTextContent(currentRow.quantityGranted.toString());
+            });
+        });
+    });
+
+    test("Check all button works correctly", () => {
```
Comment (EthanHugh) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx]: I'm having two main issues, first i don't know how to actually interact with the checkbox as nothing i do seems to change anything, also I am not too sure how to check whether it is checked, using .toBeChecked() was not working properly and so I looked at other code and they look for a tag called "Mui-checked", which I am not sure is the best way to do this.

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+                                                    >
+                                                        Qty granted
+                                                    </TableCell>
+                                                    <TableCell
+                                                        className={`${styles.width1} ${styles.noWrap}`}
+                                                    >
+                                                        <Checkbox
+                                                            color="primary"
+                                                            data-testid={`checkall-${pendingOrder.id}`}
+                                                            onChange={(e) => {
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx:134]: When this checkbox is checked, it should also be able to uncheck all the checkboxes

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+                expect(
+                    within(
+                        getByTestId(`table-${currentOrder.id}-${currentRow.id}`)
+                    ).getByText(`${mockHardware[currentRow.id - 1].manufacturer}`)
+                ).toBeInTheDocument();
+            });
+        });
+    });
+
+    test("All button changes the value correctly", () => {
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx]: try to avoid names like "does this correctly" for tests (I am a culprit of this too) but a new person would not know what "correctly" means. Maybe "All button fills dropdown to maximum value" is a better name. same for the test below.

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx (hunk 20 above)
Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx:134]: also, please use forEach here instead of map

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx (hunk 19 above)
Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.test.tsx]: Yes, this is a tricky one, the key is to add `querySelector()` to get the actual checkbox element. You also need to use an act() block because it's Formik. I only tested this for 1 table since all the tables work the same, but if you want to test all of them you should use a regular for loop instead of .forEach() because the act() calls don't work well with .forEach functions.

You also need a test to test that unchecking works.

```suggestion
    test("Check all button works correctly", () => {
                const { getByTestId } = render(<TeamPendingOrderTable />);
        const currentOrder = mockPendingOrdersInTable[0]
        const checkallBox = getByTestId(`checkall-${currentOrder.id}`).querySelector('input[type="checkbox"]');

        if (checkallBox) {
            await act(async () => {
                fireEvent.click(checkallBox);
                await waitFor(() => {
                    expect(checkallBox).toBeChecked()
                    currentOrder.hardwareInTableRow.forEach((currentRow) => {
                        expect(
                            within(getByTestId(`table-${currentOrder.id}-${currentRow.id}`))
                                .getByTestId(`${currentRow.id}-checkbox`)
                                .querySelector('input[type="checkbox"]')
                        ).toBeChecked();
                    });
                });
            })
        }
    }
```

**Review by YuYing-Liang — COMMENTED** [review]
Left some comments to help with the tests, lmk if you have any more questions!

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+                                                                if (e.target.checked) {
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx]: @EthanHugh I feel like you could minimize some logic here:
```
pendingOrder.hardwareInTableRow.forEach((row) => { props.setFieldValue(<checkbox key>, e.target.checked})
```
instead of needing 2 loops and an if statement

Other than that, your code & tests look good!

**Review by YuYing-Liang — COMMENTED** [review]
Tests look good, I left 1 comment on code style and logic.

File: hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx (hunk 22 above)
Comment (EthanHugh) [reply hackathon_site/dashboard/frontend/src/components/teamDetail/TeamPendingOrderTable/TeamPendingOrderTable.tsx]: ohh that makes sense i think i had that if statement because i didnt have that uncheck part before

**Review by YuYing-Liang — APPROVED** [review]
Looks good to me 😀
